sway/input/text_input.c

```diff
@@ -255,6 +267,253 @@ static void relay_handle_text_input(struct wl_listener *listener,
 	sway_text_input_create(relay, wlr_text_input);
 }
 
+static struct sway_layer_surface *loop_layer_surface(struct wl_list children,
+		struct wlr_layer_surface_v1 *layer)
+{
+	struct wlr_scene_node *node;
+	wl_list_for_each (node, &children, link) {
+		struct sway_layer_surface *surface = scene_descriptor_try_get(node,
+			SWAY_SCENE_DESC_LAYER_SHELL);
+		if (!surface) {
+			continue;
+		}
+		if (surface->layer_surface == layer) {
+			return surface;
+		}
+	}
+	return NULL;
+}
+
+static struct sway_layer_surface *find_layer_by_surface(
+		struct wlr_layer_surface_v1 *layer) {
```

Comment:
All of this logic to find a `struct sway_layer_surface *` from a `struct wlr_layer_surface_v1 *layer` is a bit cumbersome… Can we use the `data` field instead?

- In `sway_layer_surface_create()`, add `surface->layer_surface->data = surface`.
- In `handle_node_destroy()`, add `layer->layer_surface->data = NULL`.
- Replace `find_layer_by_surface()` usage with `struct sway_layer_surface *layer = layer_surface->data`.

Reply:
Emm. before I try to use data, but in `data`, it seems is not `sway_layer_surface`, so I use this way instead

Reply:
Oh, I see, I will  take a try

Reply:
Thank you very much